PowershellTools.HostService/ServiceManagement/Debugging/PowershellServiceHost.cs

```diff
@@ -142,38 +149,126 @@ public override System.Globalization.CultureInfo CurrentUICulture
             get { return this.originalUICultureInfo; }
         }
 
+        #region IHostSupportsInteractiveSession
 
         public bool IsRunspacePushed
         {
-            get { throw new NotImplementedException(); }
-
-            //get { return _runspaceRef.IsRunspaceOverridden; }
+            get { return _pushedRunspace != null; }
         }
 
         public void PopRunspace()
         {
-            throw new NotImplementedException();
-            
-            //UnregisterRemoteFileOpenEvent(Runspace);
-            //_runspaceRef.Revert();
+            UnregisterRemoteFileOpenEvent(Runspace);
+            Runspace = _pushedRunspace;
+            _pushedRunspace = null;
         }
 
 
         public void PushRunspace(System.Management.Automation.Runspaces.Runspace runspace)
         {
-            throw new NotImplementedException();
+            _pushedRunspace = Runspace;
+            Runspace = runspace;
+            Runspace.Debugger.SetDebugMode(DebugModes.RemoteScript);
+
+            RegisterRemoteFileOpenEvent(runspace);
+        }
+
+        Runspace IHostSupportsInteractiveSession.Runspace
+        {
+            get { return PowershellDebuggingService.Runspace; }
+        }
+
+        #endregion IHostSupportsInteractiveSession
+
+        #region private helpers
+
+        /// <summary>
+        /// Unused at the moment. Will be used for remote debugging scripts.
+        /// </summary>
+        /// <param name="remoteRunspace"></param>
+        public void RegisterRemoteFileOpenEvent(Runspace remoteRunspace)
+        {
+            remoteRunspace.Events.ReceivedEvents.PSEventReceived += new PSEventReceivedEventHandler(this.HandleRemoteSessionForwardedEvent);
+            if (remoteRunspace.RunspaceStateInfo.State != RunspaceState.Opened || remoteRunspace.RunspaceAvailability != RunspaceAvailability.Available)
+            {
+                return;
+            }
+            using (PowerShell powerShell = PowerShell.Create())
+            {
+                powerShell.Runspace = remoteRunspace;
+                powerShell.AddScript(@"
+param (

```

Comment:
Can we put these scripts-in-code in their own const file, where they're less prone to accidental breakage?  [Do we want UnitTests around them to make sure they dont' break if we ever need to tweak 'em?]